File: src/process.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

#include "process.h"
#include "macros.h"
#include "list.h"
#include "memory.h"

#define STR_BUFFER_SIZE 250

/* instructions */
typedef struct {
    INSTR_ENUM      type;
    int             cycles;
    list            memory;
} instruction;

instruction *ins_init(INSTR_ENUM t, int c, int mem_size) {
    instruction *ins = malloc(sizeof(instruction*));
    ins->type = t;
    ins->cycles = c;
    ins->memory = mem_allocate(mem_size);

    return ins;
}


/* processes */
struct pcb_struct {
    int             pc;     // program counter
    instruction**   text;   // list of instructions
    int             text_len;
    int             memory; // memory allocated
    char*           name;   
    int             cycles_left; // cycles left in the quantum
    int             time;   // how long it's been running
    int             priority;
    unsigned int    pid;
};

static char *INSTR_NAME_TABLE[] = { 
                                    "EXE", 
                                    "CALCULATE",
                                    "IO",
                                    "YIELD",
                                    "OUT"
                                  };

process pr_init(char *filename, unsigned int pid, int *memory) {
    int i, num_instr;
    byte a[3], buffer[STR_BUFFER_SIZE];
    FILE *fp = fopen(filename, "rb");
    *memory = 0;

    process p = malloc(sizeof(struct pcb_struct));
    p->text = NULL; p->name = NULL;
    p->text_len = 0;

    TRY(fread(a, 1, 2, fp) == 2);  // sanity check: first two
    TRY(a[0] == 'p' && a[1] == 'f'); // bytes should be 'pf'

    for (i = 0; i < STR_BUFFER_SIZE; i++) { // filename
        TRY(fread(a, 1, 1, fp) == 1);
        if ((buffer[i] = a[0]) == '\0')
            break;
    }
    p->name = malloc(sizeof(char)*(i+1));
    strcpy(p->name, buffer);

    TRY(fread(a, 1, 2, fp) == 2);    // memory requirements
    p->memory = (a[0] << 8) | a[1];
    if (!mem_check(p->memory)) {
        *memory = p->memory;
        TRY(THROW);
    }
    
    TRY(fread(a, 1, 1, fp) == 1);    // instruction count
    i = a[0];
    p->text = malloc(sizeof(instruction*) * i);
    for (p->text_len = 0; p->text_len < i; p->text_len++) {
        TRY(fread(a, 1, 1, fp) == 1);
        if (a[0] == CALCULATE || a[0] == IO) {
            TRY(fread(a+1, 1, 1, fp) == 1);
            if (a[1] == 255)
                a[1] = (rand() % 5) + 10; // random value
        }
        else
            a[1] = 1;
        TRY(fread(a+2, 1, 1, fp) == 1);
        p->text[p->text_len] = ins_init(a[0], a[1], a[2]);        
    }
    fclose(fp);

    p->pc = p->priority = p->time = 0;
    p->cycles_left = p->text[0]->cycles;
    p->pid = pid;
    DEBUG_PRINT("[Process] Successfully opened %s (PID %d)!", filename, p->pid);
    return p;

    // FILE CLEANUP ON ERROR
    CATCH (
        fclose(fp);
        pr_terminate(p);
        if (*memory)
            printf("[Process] Error opening %s\n", filename);
        else
            printf("[Process] Not enough memory for %s\n", filename);
        return NULL;
    )
}

void pr_terminate(process p) {
    DEBUG_PRINT("[Process] Terminating PID %d.", p->pid);
    for (int i = 0; i < p->text_len; i++) {
        mem_deallocate(p->text[i]->memory);
        l_free(p->text[i]->memory);
        free(p->text[i]);
    }
    free(p->text);
    free(p->name);
    free(p);
}

void pr_setTime(process p, int time) {
    p->cycles_left = time;
}

bool pr_run(process p) {
    p->cycles_left--;
    p->time++;
    mem_access(p->text[p->pc]->memory);
    if (p->cycles_left == 0) {
        DEBUG_PRINT("[Process] PID %d just ended %s.", p->pid, INSTR_NAME_TABLE[pr_getCurrentInstr(p)]);
        p->cycles_left = p->text[++p->pc]->cycles;
        return false;
    }
    else {
        DEBUG_PRINT("[Process] PID %d just ran %s.", p->pid, INSTR_NAME_TABLE[pr_getCurrentInstr(p)]);
        return true; 
    }
}

INSTR_ENUM pr_getCurrentInstr(process p) {
    return p->text[p->pc]->type;
}

void pr_incrementPC(process p) {
    p->cycles_left = p->text[++p->pc]->cycles;
}

void pr_print(process p, FILE *fp, bool header) {
    if (header)
        fprintf(fp, "-------------\n");
    fprintf(fp, "Process ID %d:\n", p->pid);
    fprintf(fp, "\tName: %s\n", p->name);
    fprintf(fp, "\tMemory: %d bytes\n", p->memory);
    fprintf(fp, "\tElapsed time: %d cycles\n", p->time);
    fprintf(fp, "\tText section: %d instructions\n", p->text_len);
    fprintf(fp, "\tCurrent instruction: %s\n", INSTR_NAME_TABLE[p->text[p->pc]->type]);
    fprintf(fp, "\tTime left in current instruction: %d cycles\n", p->cycles_left);
    if (header)
        fprintf(fp, "-------------\n");
}

File: src/process.py
import json

from random import randint
from threading import Lock

import src.exceptions as ex

from src.globals import DebugPrint
from src.memory import memory

class Instruction:
    def __init__(self, owner, instr):
        self.owner = owner
        self.mem_list = None    # suppresses an error
        try:
            self.op = instr["operation"]
            self.arg = {
                "acquire"   :   lambda x: x["monitor"],
                "release"   :   lambda x: x["monitor"], 
                "io"        :   lambda x: None,
                "calculate" :   lambda x: None,
                "out"       :   lambda x: None,
                "yield"     :   lambda x: None,
                "exe"       :   lambda x: None,
                "fork"      :   lambda x: None,
                "thread"    :   lambda x: x["num"],
                "send"      :   lambda x: (x["val"], x["monitor"]), 
                "read"      :   lambda x: x["monitor"]
            }[self.op](instr)
            self.cycle = {
                "acquire"   :   lambda x: 1,
                "release"   :   lambda x: 1, 
                "io"        :   lambda x: x["cycles"],
                "calculate" :   lambda x: x["cycles"],
                "out"       :   lambda x: 1,
                "yield"     :   lambda x: 1,
                "exe"       :   lambda x: 1,
                "fork"      :   lambda x: 1,
                "thread"    :   lambda x: 1,
                "send"      :   lambda x: 1, 
                "read"      :   lambda x: 1
            }[self.op](instr)
            if self.cycle is None:
                self.cycle = randint(1, 100)

            self.mem_list = memory.allocate(instr["memory"])
        except IndexError:
            DebugPrint("[Process] Improper instructions")
            raise ex.ImproperInstructionError

    def __del__(self):
        if self.mem_list:
            memory.deallocate(self.mem_list)

    def run(self):
        memory.access(self.mem_list)

class PThread:
    def __init__(self, parent, tid, text, queue, priority, daemon=False):
        self.parent = parent
        self.tid = tid
        self.text = text
        self.queue = queue
        self.priority = priority
        self.daemon = daemon

        self.pc = self.time = 0
        self.mutexes = list()
        self.cycles_left = self.text[0].cycle

    def getName(self):
        return self.parent.name

    def createThread(self, queue):
        return self.parent.createThread(self, queue)

    def setQueue(self, queue):
        self.queue = queue

    def setTime(self, time):
        self.cycles_left = time

    def run(self):
        self.time += 1
        self.cycles_left -= 1
        self.text[self.pc].run
        if self.cycles_left:
            DebugPrint(f"[Process] TID {self.tid} PID {self.parent.pid} just ran {self.text[self.pc].op}")
            return True
        else:
            DebugPrint(f"[Process] TID {self.tid} PID {self.parent.pid} just ended {self.text[self.pc].op}")
            self.pc += 1
            self.cycles_left = self.text[self.pc].cycle
            return False

    def currentInstr(self):
        return self.text[self.pc].op
        
    def hasMutex(self, mon):
        return mon in self.mutexes

    def setMutex(self, mon):
        self.mutexes.append(mon)

    def unsetMutex(self, mon):
        self.mutexes.remove(mon)

    def getInstrArg(self):
        return self.text[self.pc].arg

    def incrementPC(self):
        self.pc += 1
        self.cycles_left = self.text[self.pc].cycle

    def fork(self, pid, queue):
        return self.parent.fork(self, pid, queue)

    def __repr__(self):
        return f"TID {self.tid}/PID {self.parent.pid}"

    def __str__(self):
        s = f"Thread ID {self.tid} (Parent ID {self.parent.pid})\n"
        s += f"\tName: {self.parent.name}\n"
        s += f"\tPriority: {self.priority}\n"
        s += f"\tMemory: {self.parent.size} MB\n"
        s += f"\tElapsed time: {self.time} cycles\n"
        s += f"\tMutexes: {self.mutexes}\n"
        s += f"\tText section: {len(self.text)} instructions\n"
        s += f"\tCurrent instruction: {self.currentInstr()}\n"
        s += f"\tTime left in current instruction: {self.cycles_left}\n"

        return s    

class Process:
    def __init__(self, filename, pid, queue, fork_dump=None):
        if fork_dump:
            dump = fork_dump
        else:
            with open(filename, "r") as f:
                dump = json.load(f)

        self.name = dump["name"]
        self.priority = dump["priority"]
        self.daemon = dump["daemon"] if "daemon" in dump else False

        self.size = 0
        for instr in dump["instructions"]:
            self.size += instr["memory"]
        if self.size > memory.remaining():
            raise ex.MemoryAllocationError

        self.text = list()
        for instr in dump["instructions"]:
            self.text.append(Instruction(self, instr))

        self.thread_text = list()
        if "threadinstr" in dump:
            for instr in dump["threadinstr"]:
                self.thread_text.append(Instruction(self, instr))

        self.next_tid = 0
        self.threads = [PThread(self, 0, self.text, queue, self.priority, daemon=self.daemon)]
        self.pid = pid
        self.dump = dump
        self.lock = Lock()

    def getMain(self):
        return self.threads[0]

    def createThread(self, requester, queue):
        with self.lock:
            self.next_tid += 1
            thr = PThread(self, self.next_tid, requester.text[requester.pc:], queue, self.priority, daemon=self.daemon)
            self.threads.append(thr)
            ret = self.threads[-1]
        return ret

    def setQueue(self, queue):
        self.queue = queue

    def fork(self, requester, pid, queue):
        new_proc = Process("", pid, queue, fork_dump=self.dump)
        main = new_proc.getMain()
        main.pc = requester.pc 
        main.cycles_left = requester.cycles_left

        return new_proc.getMain()

    # def setTime(self, time):
    #     self.cycles_left = time

    # def currentInstr(self):
    #     return self.text[self.pc].op

    # def getInstrArg(self):
    #     return self.text[self.pc].arg

    # def incrementPC(self):
    #     self.pc += 1
    #     self.cycles_left = self.text[self.pc].cycle

    

    